Classes/common/HTTP/CDTURLSessionInterceptorContext.h

```diff
@@ -0,0 +1,39 @@
+//
+//  CDTURLSEssionFilterContext.h
+//  
+//
+//  Created by Rhys Short on 17/08/2015.
+//
+//
+
+#import <Foundation/Foundation.h>
+#import "CDTMacros.h"
+
+@interface CDTURLSessionInterceptorContext : NSObject
+
+@property (nonnull, readonly, nonatomic, strong) NSMutableURLRequest *request;
+
+- (nullable instancetype) init UNAVAILABLE_ATTRIBUTE;
+
+/**
+ *  Initalizes a CDTURLSessionInterceptorContext
+ *
+ *  @param request the request this context should represent
+ *
+ **/
+- (nullable instancetype) initWithRequest:(nonnull NSMutableURLRequest*)request NS_DESIGNATED_INITIALIZER;
+
+
+@end
+
+@protocol CDTURLSessionInterceptor <NSObject>
```

Comment:
I thought I started a discussion about this as part of the discussion here -> https://github.com/cloudant/CDTDatastore/pull/186#discussion-diff-37550563. 
 I thought since we can have optional method implementations in Objc easier than with Java, having both, in the same protocol makes it easier (and a little neater) for developers if they want to implement something which intercepts both requests and responses.